cuda: Add Windows Support

matheusd opened the pull request:
This adds support to building the CUDA-enabled version of `gominer` on Windows, as well as adding instructions on how to do so.

Additionally, it switches from using a Makefile to using `go generate` in order to make the building process more idiomatic to Go projects.

matheusd commented on README.md:
```diff
@@ -193,8 +193,16 @@ The following sections provide instructions for these combinations:
 
-**Build instructions are not available yet.  They will be provided at a later
-date**.
+Use the same instructions for fetching and installing both the NVIDIA driver,
+build tools and a copy of `gominer`'s source code.
```
Instead of adding a reference here, maybe I should generalize the generic bits of the instructions for both OpenCL and CUDA bits at an earlier section?

davecgh replied:
Yeah, it probably makes sense to split it as amd/nvidia, then have a common nvidia prereqs section and break out the nvidia opencl / cuda specific bits.

matheusd commented on README.md:
```diff
@@ -245,8 +253,32 @@ date**.
+Toolkit version 12.2. If the automatic builder for `gominer` does not work on
+your system, you many need to manually setup the various tools.
+
+Use the same instructions for fetching and installing MSYS2, build tools and a
+fetch a copy of `gominer`'s source code as the OpenCL build.
```
Same as other comment

davecgh commented on README.md:
```diff
@@ -267,4 +299,4 @@ NVIDIA Tesla V100S     | 14.6 Gh/s
 NVIDIA RTX 4070        | 14.9 Gh/s
 NVIDIA RTX 3080        | 15.2 Gh/s
 NVIDIA RTX 3090        | 17.6 Gh/s
-AMD 7900 XTX           | 27.2 Gh/s
```
It's not directly related, but since you're updating the readme already, mind changing this to 23.8?  The user updated it after running for a while and it stabilized.

davecgh commented on cuda_builder.go:
```diff
@@ -0,0 +1,282 @@
+		"-I.", "blake3.cu", "-o", "blake3.dll")
+}
+
+// buildBlake3Windows builds the blake3.a library with a compiled version of
```
```suggestion
buildBlake3Default builds the blake3.a library with a compiled version of
```

davecgh reviewed (commented):
This looks great overall.  The cuda building parts are a very nice addition.

I wonder if the generate portion should just handle the building of the compiled blake3 cuda kernel and then a normal `go build -tags cuda` to actually build gominer after that?

matheusd commented:
> I wonder if the generate portion should just handle the building of the compiled blake3 cuda kernel and then a normal go build -tags cuda to actually build gominer after that?

Then we're back at having the user setup their own `CGO_CFLAGS` and `CGO_LDFLAGS` by hand, because those are also needed during the `go build` stage. Adding `#cgo` directives to the source code is not sufficient, because `barnex/cuda5` requires `cuda.h`, and that isn't available by default on windows without setting up the env vars.

davecgh commented:
> Then we're back at having the user setup their own `CGO_CFLAGS` and `CGO_LDFLAGS` by hand, because those are also needed during the `go build` stage. Adding `#cgo` directives to the source code is not sufficient, because `barnex/cuda5` requires `cuda.h`, and that isn't available by default on windows without setting up the env vars.

Ah, alright.  I thought it was only needed for the nvcc bits to build the kernel.

matheusd commented:
Updated readme with the various sections ([view compiled file](https://github.com/decred/gominer/blob/09c22f072069b4abaed9a35f95fc7e41ce87429b/README.md)).

Also, I added a document which describes how to manually setup CUDA builds in windows ([view](https://github.com/decred/gominer/blob/09c22f072069b4abaed9a35f95fc7e41ce87429b/docs/cuda-manual-windows-build.md))

davecgh commented on README.md:
```diff
@@ -209,15 +256,26 @@ date**.
   - `wget https://github.com/GPUOpen-LibrariesAndSDKs/OCL-SDK/files/1406216/lightOCLSDK.zip`
   - `unzip -d /c/appsdk lightOCLSDK.zip`
-  - `git clone https://github.com/decred/gominer`
 - **Close the `MSYS2 MINGW64` shell and relaunch it**
```
I'm fairly certain you'll need this up in the prereqs section for both opencl and cuda (prior to the "Jump to the appropriate...").  It's really for the gcc and Go env variables.  Unzipping the opencl sdk doesn't need the shell closed and reopened.

davecgh commented on README.md:
```diff
@@ -243,10 +301,41 @@ date**.
-**NOTE**: The CUDA version of the `gominer` is not yet compatible with
-Windows.
+After creating an appropriate `gominer` binary (`gominer.exe` on Windows), you
+may verify whether it is properly working with the following procedure:
```
There is nothing after this, but the text indicates there should be.

davecgh commented on README.md:
```diff
@@ -178,29 +206,48 @@ The following sections provide instructions for these combinations:
+
+###### OpenCL on Debian
+
+- Install the OpenCL headers, OpenCL Installable Client driver, OpenCL lib
```
```suggestion
- Install the OpenCL headers, OpenCL Installable Client driver, and OpenCL lib
```

davecgh commented on README.md:
```diff
@@ -178,29 +206,48 @@ The following sections provide instructions for these combinations:
 
-**Build instructions are not available yet.  They will be provided at a later
-date**.
+###### CUDA on Debian
```
This section seems off in its placement.  It's a subsection of "AMD on Debian Bookworm".

matheusd replied:
Do you think I should drop this (and leave CUDA only on the Ubuntu section) or rename the parent to just "Debian Bookwork"?

davecgh replied:
Yeah, probably just making it Debian Bookworm and then a slight rearrangement to refactor out the common bits and then something like:

* Jump to the appropriate section for either [OpenCL with AMD](#link), [OpenCL with NVIDIA](#link), or [CUDA with NVIDIA](#link) depending on which type of GPU you have and the desired tookit

davecgh commented on docs/cuda-manual-windows-build.md:
```diff
@@ -0,0 +1,153 @@
+Building the CUDA-enabled `gominer` on a Windows platform is tricky, requires
+several GB worth of downloads and the [automatic builder
+script](/cuda_builder.go) that is called when running `go generate -tags cuda`
+is not guaranteed to work on every combination of setup.
```
```suggestion
is not guaranteed to work on every combination of setups.
```

davecgh commented on docs/cuda-manual-windows-build.md:
```diff
@@ -0,0 +1,153 @@
+  in the user's `$PATH` environment variable.
+- A default installation of MSYS2 ignores the global and user's `$PATH` anyway.
+- Having files in `c:\Program Files\` (path with a space) causes all sorts of
+  issues in all layer of the compilation stack.  In particular, [cgo directives
```
```suggestion
  issues in all layers of the compilation stack.  In particular, [cgo directives
```

davecgh commented on docs/cuda-manual-windows-build.md:
```diff
@@ -0,0 +1,153 @@
+  `gcc`, `git`, `go`, `unzip`:
+  - `pacman -S mingw-w64-x86_64-gcc mingw-w64-x86_64-tools mingw-w64-x86_64-go git unzip`
+  - `git clone https://github.com/decred/gominer`
+
```
The shell will have to be restarted here to pick up the gcc and go env variables properly.

davecgh commented on docs/cuda-manual-windows-build.md:
```diff
@@ -0,0 +1,153 @@
+  - This is generally `c:\Program Files\NVIDIA GPU Toolkit\CUDA\vXX.Y`
+- Create a junction for the target directory from within MSYS2:
+  - `cd gominer`
+  - `cmd.exe //c mklink /J nvidia "c:\Program Files\NVIDIA GPU Toolkit\CUDA\vXX.Y"`
```
I think it has to be a double slash on the J so the msys2 cmd translation layer doesn't eat it.

```suggestion
  - `cmd.exe //c mklink //J nvidia "c:\Program Files\NVIDIA GPU Toolkit\CUDA\vXX.Y"`
```

davecgh commented on docs/cuda-manual-windows-build.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+Restart the MSYS2 shell after editing the file. To know whether the values are
+being used, issue `echo` commands (for example, `echo "$CGO_CFLAGS"`).
````
```suggestion
being used, issue `echo` commands (for example, `echo $CGO_CFLAGS`).
```

davecgh commented on docs/cuda-manual-windows-build.md:
````diff
@@ -0,0 +1,153 @@
+$ ./gominer -l
+```
+
+Continue with the standard [configure and run](#configuring-gominer) procedure.
````
```suggestion
Continue with the standard [configure and run](#../configuring-gominer) procedure.
```

matheusd commented:
Made a first batch of review fixes.

matheusd commented:
Updated with changes to the debian section.

davecgh commented:
I'll merge this after the squash.
